File: quration-core/examples/trotter.cpp
#include <boost/program_options.hpp>
#include <nlohmann/json.hpp>

#include <cerrno>
#include <cstddef>
#include <cstdint>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <map>
#include <stdexcept>
#include <string>
#include <vector>

#include "qret/frontend/argument.h"
#include "qret/frontend/builder.h"
#include "qret/frontend/circuit_generator.h"
#include "qret/frontend/intrinsic.h"
#include "qret/ir/instruction.h"
#include "qret/ir/json.h"
#include "qret/transforms/ipo/inliner.h"

using json = nlohmann::json;
using namespace qret;
using frontend::CircuitBuilder, frontend::CircuitGenerator;

enum class PauliChars : char { X = 'X', Y = 'Y', Z = 'Z' };
std::string PauliOpToStr(const PauliChars op) {
    switch (op) {
        case PauliChars::X:
            return "X";
        case PauliChars::Y:
            return "Y";
        case PauliChars::Z:
            return "Z";
        default: {
            throw std::runtime_error("Invalid PauliOp");
        }
    }
}
struct PauliTerm {
    double coeff;
    std::map<std::size_t, PauliChars> pauli_string;

    PauliTerm() = default;
    explicit PauliTerm(double coeff, std::map<std::size_t, PauliChars>& pauli_string)
        : coeff(coeff)
        , pauli_string(pauli_string) {

        };
};
void from_json(const json& j, PauliTerm& pauli_term) {
    j.at("coeff").get_to(pauli_term.coeff);

    pauli_term.pauli_string.clear();
    for (const auto& item : j.at("pauli_string").items()) {
        const auto qubit_index = static_cast<std::size_t>(std::stoull(item.key()));

        const auto s = item.value().get_ref<const std::string&>();
        pauli_term.pauli_string[qubit_index] = static_cast<PauliChars>(s[0]);
    }
}
void to_json(json& j, const PauliTerm& pauli_term) {
    json pauli_string_json;
    for (const auto& [qubit_index, pauli_op] : pauli_term.pauli_string) {
        pauli_string_json[std::to_string(qubit_index)] = PauliOpToStr(pauli_op);
    }

    j["coeff"] = pauli_term.coeff;
    j["pauli_string"] = pauli_string_json;
}

struct Hamiltonian {
    std::size_t num_qubits;
    std::vector<PauliTerm> pauli_terms;

    Hamiltonian() = default;
    explicit Hamiltonian(std::size_t num_qubits, std::vector<PauliTerm>& pauli_terms)
        : num_qubits(num_qubits)
        , pauli_terms(pauli_terms) {};
};
void from_json(const json& j, Hamiltonian& h) {
    j.at("num_qubits").get_to(h.num_qubits);
    j.at("pauli_terms").get_to(h.pauli_terms);
}
void to_json(json& j, const Hamiltonian& h) {
    j["num_qubits"] = h.num_qubits;
    j["pauli_terms"] = h.pauli_terms;
}
Hamiltonian ReadHamiltonian(const std::string& path) {
    auto ifs = std::ifstream(path.data());
    if (!ifs.good()) {
        throw std::runtime_error("Could not open file: " + path);
    }
    json j;
    ifs >> j;
    return j.get<Hamiltonian>();
}
struct PauliRotationGen : CircuitGenerator {
    std::size_t num_qubits;
    std::map<std::size_t, PauliChars> pauli_string;
    double angle;

    explicit PauliRotationGen(
            CircuitBuilder* builder,
            const std::size_t num_qubits,
            const std::map<std::size_t, PauliChars>& pauli_string,
            const double angle
    )
        : CircuitGenerator(builder)
        , num_qubits(num_qubits)
        , pauli_string(pauli_string)
        , angle(angle) {}
    std::string GetName() const override {
        auto serialized_pauli_string = std::string();
        for (const auto& pair : pauli_string) {
            serialized_pauli_string += fmt::format("{}{}", pair.first, PauliOpToStr(pair.second));
        }
        return fmt::format("pauli_rotation({}, {:.4f})", serialized_pauli_string, angle);
    }
    std::string GetCacheKey() const override {
        return GetName();
    }
    void SetArgument(Argument& arg) const override {
        arg.Add("q", Type::Qubit, num_qubits, Attribute::Operate);
    }
    frontend::Circuit* Generate() const override {
        if (IsCached()) {
            return GetCachedCircuit();
        }
        BeginCircuitDefinition();
        auto q = GetQubits(0);

        for (const auto& [idx, op] : pauli_string) {
            switch (op) {
                case PauliChars::X:
                    frontend::gate::H(q[idx]);
                    break;
                case PauliChars::Y:
                    frontend::gate::SDag(q[idx]);
                    frontend::gate::H(q[idx]);
                    break;
                case PauliChars::Z:
                    break;
                default:
                    throw std::runtime_error("Unsupported pauli character.");
            }
        }

        auto target_qubits = std::vector<std::size_t>();
        for (const auto& [idx, _] : pauli_string) {
            target_qubits.emplace_back(idx);
        }
        for (auto i = std::size_t{0}; i < target_qubits.size() - 1; ++i) {
            frontend::gate::CX(q[target_qubits[i]], q[target_qubits[i + 1]]);
        }
        frontend::gate::RZ(q[target_qubits.back()], angle);

        for (const auto& [idx, op] : pauli_string) {
            switch (op) {
                case PauliChars::X:
                    frontend::gate::H(q[idx]);
                    break;
                case PauliChars::Y:
                    frontend::gate::H(q[idx]);
                    frontend::gate::S(q[idx]);
                    break;
                case PauliChars::Z:
                    break;
                default:
                    throw std::runtime_error("Unsupported pauli character.");
            }
        }

        return EndCircuitDefinition();
    }
};
void AddPauliRotation(
        const frontend::Qubits& q,
        const std::map<std::size_t, PauliChars>& pauli_string,
        const double angle
) {
    auto gen = PauliRotationGen(q.GetBuilder(), q.Size(), pauli_string, angle);
    auto* circuit = gen.Generate();
    (*circuit)(q);
}
struct TrotterStepGen : CircuitGenerator {
    Hamiltonian hamiltonian;
    double step_factor;
    static inline const char* Name = "trotter_step";

    explicit TrotterStepGen(CircuitBuilder* builder, const Hamiltonian& h, const double step_factor)
        : CircuitGenerator(builder)
        , hamiltonian(h)
        , step_factor(step_factor) {}
    std::string GetName() const override {
        return Name;
    }
    std::string GetCacheKey() const override {
        return GetName();
    }
    void SetArgument(Argument& arg) const override {
        arg.Add("q", Type::Qubit, hamiltonian.num_qubits, Attribute::Operate);
    }
    frontend::Circuit* Generate() const override {
        if (IsCached()) {
            return GetCachedCircuit();
        }

        BeginCircuitDefinition();
        auto q = GetQubits(0);
        for (const auto& term : hamiltonian.pauli_terms) {
            const auto step_angle = 2 * term.coeff * step_factor;
            AddPauliRotation(q, term.pauli_string, step_angle);
        }
        return EndCircuitDefinition();
    }
};
void AddTrotterStep(
        const frontend::Qubits& q,
        const Hamiltonian& hamiltonian,
        const double step_factor
) {
    auto gen = TrotterStepGen(q.GetBuilder(), hamiltonian, step_factor);
    auto* circuit = gen.Generate();
    (*circuit)(q);
}
struct TrotterCircuitGen : CircuitGenerator {
    Hamiltonian hamiltonian;
    std::size_t num_trotter_steps;
    double time;

    TrotterCircuitGen(
            CircuitBuilder* builder,
            const Hamiltonian& h,
            const std::size_t num_trotter_steps,
            const double time
    )
        : CircuitGenerator(builder)
        , hamiltonian(h)
        , num_trotter_steps(num_trotter_steps)
        , time(time) {}
    std::string GetName() const override {
        return fmt::format("trotter_circuit({})", num_trotter_steps);
    }
    std::string GetCacheKey() const override {
        return GetName();
    }
    void SetArgument(Argument& arg) const override {
        arg.Add("q", Type::Qubit, hamiltonian.num_qubits, Attribute::Operate);
    }
    frontend::Circuit* Generate() const override {
        if (IsCached()) {
            return GetCachedCircuit();
        }
        BeginCircuitDefinition();
        auto q = GetQubits(0);
        for (auto i = std::size_t{0}; i < num_trotter_steps; ++i) {
            AddTrotterStep(q, hamiltonian, time / num_trotter_steps);
        }
        return EndCircuitDefinition();
    }
};

std::int32_t main(std::int32_t argc, const char* const* const argv) {
    std::filesystem::path source_file_path = __FILE__;
    const auto default_input_file =
            source_file_path.parent_path() / "data/trotter/1d_ising_hamiltonian.json";
    namespace po = boost::program_options;
    po::options_description desc("Simulation of Trotter decomposition");
    desc.add_options()
        ("help", "Print usage instructions")
        ("file", po::value<std::string>()->default_value(default_input_file.string()), "Path to JSON file of input Hamiltonian")
        ("time", po::value<double>()->default_value(1.0), "Time for time evolution")
        ("num_trotter_steps", po::value<std::size_t>()->default_value(1), "Number of trotter steps")
        ("out", po::value<std::string>()->default_value("out.json"), "Path to the output file")
        ("inline", "Option to enable inline expansion for all modules in the Trotter circuit.");

    po::variables_map vm;
    store(parse_command_line(argc, argv, desc), vm);
    notify(vm);

    if (vm.count("help") > 0) {
        std::cout << desc << std::endl;
        return 0;
    }

    std::string input_file;
    if (vm.count("file") > 0) {
        input_file = vm["file"].as<std::string>();
    }
    const auto h = ReadHamiltonian(input_file);

    const auto time = vm["time"].as<double>();
    const auto num_trotter_steps = vm["num_trotter_steps"].as<std::size_t>();
    ir::IRContext context;
    auto* module = ir::Module::Create("simulation_time_evolution_with_trotter", context);
    auto builder = CircuitBuilder(module);
    auto gen = TrotterCircuitGen(&builder, h, num_trotter_steps, time);
    auto* circuit = gen.Generate();
    auto* ir_circuit = circuit->GetIR();

    // Inline expansion
    if (vm.count("inline") > 0) {
        ir::RecursiveInlinerPass().RunOnFunction(*ir_circuit);
    }

    std::string output_file;
    if (vm.count("out") > 0) {
        output_file = vm["out"].as<std::string>();
    }
    std::ofstream ofs(output_file);
    if (!ofs) {
        std::cerr << "Failed to open output file: " << output_file << std::endl;
        return 1;
    }
    auto module_json = qret::Json(*module);
    ofs << module_json;
    ofs.close();

    return 0;
}

File: quration-core/python/examples/trotter.py
"""Example of pyqret: Trotter expansion."""

import argparse
import json
from pathlib import Path
from pprint import pprint
from typing import Literal

from pydantic import BaseModel, Field
from pyqret.frontend import Argument, CircuitBuilder, CircuitGenerator, Context, Module, Opcode, Qubits
from pyqret.frontend.gate import intrinsic
from pyqret.frontend.transforms import inline_call_inst

PauliChars = Literal["X", "Y", "Z"]


class PauliTerm(BaseModel):
    """A single Pauli string with a coefficient.

    Attributes:
        coeff (float): Coefficient
        pauli_string (dict[int, str]): Pauli string with qubit index as key and operator as value.
    """

    coeff: float = Field()
    pauli_string: dict[int, PauliChars] = Field(default_factory=dict)


class Hamiltonian(BaseModel):
    """Hamiltonian defined by Pauli strings.

    Attributes:
        num_qubits (int): The number of qubits in the Hamiltonian.
        pauli_terms (list[PauliTerm]): List of Pauli terms that make up the Hamiltonian.
    """

    num_qubits: int = Field(ge=1)
    pauli_terms: list[PauliTerm] = Field(default_factory=list)


def read_hamiltonian(file: Path) -> Hamiltonian:
    """Read Hamiltonian from a JSON file.

    Args:
        file (Path): Path to the JSON file.

    Returns:
        Hamiltonian: The Hamiltonian object.
    """
    with file.open(mode="r", encoding="utf-8") as f:
        json_obj = json.load(f)

    return Hamiltonian(**json_obj)


class PauliRotation(CircuitGenerator):
    def __init__(
        self,
        builder: CircuitBuilder,
        pauli_string: dict[int, PauliChars],
        angle: float,
        num_qubits: int,
    ) -> None:
        super().__init__(builder)
        self.pauli_string = pauli_string
        self.angle = angle
        self.num_qubits = num_qubits

    def name(self) -> str:
        serialized_pauli_string = "".join([
            f"{pauli_char}{qubit_idx}" for qubit_idx, pauli_char in self.pauli_string.items()
        ])
        return f"pauli_rotation({serialized_pauli_string}, {self.angle:.4f})"

    def arg(self) -> Argument:
        ret = Argument()
        ret.add_operates("q", self.num_qubits)
        return ret

    def logic(self, arg: Argument) -> None:  # noqa: C901
        q: Qubits = arg["q"]  # pyright: ignore[reportAssignmentType]
        for index, pauli_char in self.pauli_string.items():
            if pauli_char == "X":
                intrinsic.h(q[index])
            elif pauli_char == "Y":
                intrinsic.sdag(q[index])
                intrinsic.h(q[index])
            elif pauli_char == "Z":
                pass

        target_qubits = list(self.pauli_string.keys())
        for i in range(len(target_qubits) - 1):
            intrinsic.cx(q[target_qubits[i]], q[target_qubits[i + 1]])

        intrinsic.rz(q[target_qubits[-1]], self.angle)

        for i in reversed(range(len(target_qubits) - 1)):
            intrinsic.cx(q[target_qubits[i]], q[target_qubits[i + 1]])

        for index, pauli_char in self.pauli_string.items():
            if pauli_char == "X":
                intrinsic.h(q[index])
            elif pauli_char == "Y":
                intrinsic.h(q[index])
                intrinsic.s(q[index])
            elif pauli_char == "Z":
                pass


class TrotterStep(CircuitGenerator):
    def __init__(
        self,
        builder: CircuitBuilder,
        hamiltonian: Hamiltonian,
        step_factor: float,
    ) -> None:
        super().__init__(builder)
        self.hamiltonian = hamiltonian
        self.step_factor = step_factor
        self.step_factor = step_factor

    def name(self) -> str:
        return "trotter_step"

    def arg(self) -> Argument:
        ret = Argument()
        ret.add_operates("q", self.hamiltonian.num_qubits)
        return ret

    def logic(self, arg: Argument) -> None:
        q: Qubits = arg["q"]  # pyright: ignore[reportAssignmentType]

        def add_pauli_rotation(pauli_string: dict[int, PauliChars], angle: float) -> None:
            gen = PauliRotation(self.builder, pauli_string=pauli_string, angle=angle, num_qubits=q.size())
            circuit = gen.generate()
            circuit(q)

        for pauli_term in self.hamiltonian.pauli_terms:
            step_angle = 2 * pauli_term.coeff * self.step_factor
            add_pauli_rotation(pauli_term.pauli_string, step_angle)


class TrotterCircuit(CircuitGenerator):
    def __init__(self, builder: CircuitBuilder, hamiltonian: Hamiltonian, time: float, num_trotter_steps: int) -> None:
        super().__init__(builder)
        self.hamiltonian = hamiltonian
        self.time = time
        self.num_trotter_steps = num_trotter_steps

    def name(self) -> str:
        return f"trotter_circuit({self.num_trotter_steps})"

    def arg(self) -> Argument:
        ret = Argument()
        ret.add_operates("q", self.hamiltonian.num_qubits)
        return ret

    def logic(self, arg: Argument) -> None:
        q = arg["q"]

        def add_trotter_step() -> None:
            gen = TrotterStep(self.builder, self.hamiltonian, self.time / self.num_trotter_steps)
            circuit = gen.generate()
            circuit(q)

        for _ in range(self.num_trotter_steps):
            add_trotter_step()


def main() -> None:
    """Simulating Trotter decomposition of the specied Hamiltonian."""
    parser = argparse.ArgumentParser(description="Simulation of Trotter decomposition")
    parser.add_argument(
        "-f",
        "--file",
        type=str,
        required=True,
        help="Path to JSON file of input Hamiltonian",
    )
    parser.add_argument(
        "-t",
        "--time",
        type=float,
        default=1.0,
        help="Time for time evolution",
    )
    parser.add_argument(
        "-n",
        "--num_trotter_steps",
        type=int,
        default=1,
        help="Number of trotter steps",
    )
    parser.add_argument(
        "-o",
        "--out",
        type=str,
        default="out.json",
        help="Path to the output IR file",
    )
    parser.add_argument(
        "--inline",
        default=False,
        action="store_true",
        help="Option to enable inline expansion for all modules in the Trotter circuit.",
    )

    args = parser.parse_args()
    hamiltonian = read_hamiltonian(Path(args.file))

    # Create the circuit
    context = Context()
    module = Module("simulation_time_evolution_with_trotter", context)
    builder = CircuitBuilder(module)
    gen = TrotterCircuit(builder, hamiltonian, args.time, args.num_trotter_steps)
    circuit = gen.generate()

    # Inline expansion
    if args.inline:
        call_in_step = []
        step_circuit = module.get_circuit("trotter_step")
        for bb in step_circuit.get_ir():
            for inst in bb:
                if inst.get_opcode() == Opcode.Call:
                    call_in_step.append(inst)  # noqa: PERF401
        for i in range(len(call_in_step)):
            inline_call_inst(call_in_step[i])

        call_insts = []
        for bb in circuit.get_ir():
            for inst in bb:
                if inst.get_opcode() == Opcode.Call:
                    call_insts.append(inst)  # noqa: PERF401
        for i in range(len(call_insts)):
            inline_call_inst(call_insts[i])

    module.dump(args.out)


if __name__ == "__main__":
    main()
